drivers: led_strip: ws2812: Remove WS2812_STRIP_DRIVER config

soburi opened the pull request:
- Removing `choice WS2812_STRIP_DRIVER` to enable the use of multiple types of WS2812 drivers.
   Also, `menuconfig WS2812_STRIP` will be deleted as it does not correspond to the appropriate settings.

- Applying the modern way which is adding `default y` and `depends on DT_HAS_...` to enable configs.

- Unify `led_apa102` and `led_lpd8806` samples into `led_strip` and remove these.

soburi committed:
samples: drivers: Rename `led_ws2812` to `led_strip`

Rename it so that it can be used as a sample of various LED strips.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: led_strip: ws2812: Remove WS2812_STRIP and WS2812_STRIP_DRIVER

Removing `choice WS2812_STRIP_DRIVER` to enable the use of multiple
types of WS2812 drivers.
Also, `menuconfig WS2812_STRIP` will be deleted as it does not
correspond to the appropriate settings.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
samples: drivers: led_strip: Unify `led_apa102` and `led_lpd8806` samples

Unify `led_apa102` and `led_lpd8806` samples into `led_strip`
and remove these.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: led_strip: Enabling config with DT_HAS_..

Applying the modern way which is adding `default y` and
`depends on DT_HAS_...` to enable configs.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

erwango commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -7,50 +7,48 @@
+	bool "WS2812 LED strip SPI driver"
+	default y
+	depends on DT_HAS_WORLDSEMI_WS2812_SPI_ENABLED
 	depends on SPI
```
We're now allowed to use `select SPI`, this will allow some additional clean up.

soburi replied:
I got it. Add `select SPI` and remove duplicate from *.conf.

erwango commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -7,50 +7,48 @@
+	bool "WS2812 LED strip I2S driver"
+	default y
+	depends on DT_HAS_WORLDSEMI_WS2812_I2S_ENABLED
 	depends on I2S
```
ditto: `select I2S`

soburi replied:
I did the same with SPI.

erwango commented on drivers/led_strip/Kconfig.apa102:
```diff
@@ -6,6 +6,8 @@
 	bool "APA102 SPI LED strip driver"
+	default y
+	depends on DT_HAS_APA_APA102_ENABLED
 	depends on SPI
```
Ditto

soburi replied:
I did the same with SPI.

erwango commented on drivers/led_strip/Kconfig.lpd880x:
```diff
@@ -3,6 +3,8 @@
 	bool "LPD880x SPI LED strip driver"
+	default y
+	depends on DT_HAS_GREELED_LPD8803_ENABLED || DT_HAS_GREELED_LPD8806_ENABLED
 	depends on SPI
```
Ditto

soburi replied:
I did the same with SPI.

simonguinot commented on boards/vcc-gnd/yd_esp32/doc/index.rst:
```diff
@@ -233,7 +233,7 @@ RGB LED
 =======
```
There is a typo in the commit title: led_2812 instead of led_ws2812

soburi replied:
Oops!

Fixed it.

simonguinot commented on samples/drivers/led_strip/README.rst:
```diff
@@ -44,11 +44,11 @@ property. See the overlay file
 :zephyr_file:`samples/drivers/led_ws2812/boards/thingy52_nrf52832.overlay` for more detail.
```
There are still several occurrences of led_ws2812 to replace with led_string (such above) in this file ^

soburi replied:
All `led_ws2812` change to `led_strip`.

simonguinot commented on boards/vcc-gnd/yd_esp32/doc/index.rst:
```diff
@@ -233,7 +233,7 @@ RGB LED
+Here is an example of how to test it using the :zephyr:code-sample:`led-strip` application.
 
 .. zephyr-app-commands::
    :zephyr-app: samples/drivers/led_ws2812
```
Please, replace the "samples/drivers/led_ws2812" above with "samples/drivers/led_strip".

And you need to fix the `boards/shields/adafruit_neopixel_grid_bff/doc/index.rst` file as well.

soburi replied:
Same as above.

simonguinot commented on boards/adafruit/itsybitsy/doc/index.rst:
```diff
@@ -123,7 +123,7 @@ The :zephyr:code-sample:`button` sample lets you test the buttons (switches) and
 The :zephyr:code-sample:`blinky` sample lets you test the red LED.
 
 The DotStar LED has been implemented as a SPI device and can be tested
-with the :zephyr:code-sample:`led-apa102` sample application.
```
I am not sure but maybe we should mention in the release not that theses samples have been removed and replaced by the `led_strip` sample ?

soburi replied:
At a glance, It seems the removing samples are not usually mentioned in the release note.

thedjnK replied:
Not been mentioned before

simonguinot replied:
Perfect then !

soburi commented on drivers/led_strip/Kconfig.apa102:
```diff
@@ -6,6 +6,8 @@
 config APA102_STRIP
 	bool "APA102 SPI LED strip driver"
+	default y
+	depends on DT_HAS_APA_APA102_ENABLED
```
Fixed it.

erwango commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -7,27 +7,16 @@
+	select I2S
 	help
 	  Uses the I2S peripheral, memory usage is 4 bytes per color,
 	  times the number of pixels. A few more for the start and end
```
These could be removed as well if you use following example:
```
	select I2C if $(dt_compat_on_bus,$(DT_COMPAT_BOSCH_BME680),i2c)
	select SPI if $(dt_compat_on_bus,$(DT_COMPAT_BOSCH_BME680),spi)
```

See also use of `DT_ANY_INST_ON_BUS_STATUS_OKAY` in drivers/sensors

soburi replied:
Thank you for inform. I fixed it.

soburi commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -7,50 +7,48 @@
-	bool "SPI driver"
+	bool "WS2812 LED strip SPI driver"
+	default y
+	depends on DT_HAS_WORLDSEMI_WS2812_SPI_ENABLED
```
Reconsidered the division to make the commits clearer.

thedjnK commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -7,27 +7,16 @@
-	default WS2812_STRIP_SPI
-	depends on WS2812_STRIP
-
 config WS2812_STRIP_SPI
```
These will now just show "SPI driver" with no context, needs either a menu or more text

thedjnK commented on samples/drivers/led_strip/README.rst:
```diff
@@ -46,37 +62,13 @@ property. See the overlay file
 modified by changing the :kconfig:option:`CONFIG_SAMPLE_LED_UPDATE_DELAY`.
 
+If there is no chain-length property in the devicetree node,
+you need to set the number of LEDs in the :kconfig:option:`SAMPLE_LED_STRIP_LENGTH`
```
needs CONFIG_ prefix

thedjnK replied:
@gmarull why is this not failing CI?

soburi replied:
Thank you for catching the error.
Fixed it.

simonguinot commented on boards/seagate/legend/legend.dts:
```diff
@@ -22,6 +22,7 @@
 	aliases {
 		watchdog0 = &iwdg;
 		spi-flash0 = &spi_nor;
+		led-strip = &led_strip_spi;
```
I am not sure... This alias is not missing. It is provided in the `legend_*.overlay` DT files. But since it's used on all revisions of the device, I don't know why it wasn't shared in `legend.dts` in the first place...

soburi replied:
It was a half-hearted fix, so I fixed it.
Even if we do not modify this, it will not affect the build, but I think it is better to make it common.

simonguinot replied:
Please fix the name of my employer in the commit title :)

seagage -> seagate

soburi replied:
Oops, I'm very sorry about that. fixed.

thedjnK commented on samples/drivers/led_strip/Kconfig:
```diff
@@ -9,6 +9,12 @@ config SAMPLE_LED_UPDATE_DELAY
+	int "LED strip length"
+	default 0
+	help
+	  Number of LEDs in the strip
```
Needs to say that 0 means use dts

soburi replied:
Added description.

simonguinot commented on samples/drivers/led_ws2812/boards/actinius_icarus_som_dk_nrf9160.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_WS2812_STRIP_SPI=y
```
Is this a preparation for the removal of `WS2812_STRIP_DRIVER` (and `WS2812_STRIP_SPI` default) coming in the next patch of the series?

If yes, then I think you should merge these two patches.

simonguinot commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -8,36 +8,47 @@
+	depends on DT_HAS_WORLDSEMI_WS2812_SPI_ENABLED
 	select SPI if $(dt_compat_on_bus,$(DT_COMPAT_WORLDSEMI_WS2812_SPI),spi)
 	help
+	  Enable driver for WS2812 (and compatibles) LED strip that using SPI.
```
I think you can remove "that" ^

Suggestion:
`Enable driver for WS2812 (and compatibles) LED strip controllers using SPI.`
Or:
`Enable SPI driver for WS2812 (and compatibles) LED strip controllers.`

This proposition also applies to the other help messages below.

soburi replied:
I applied your suggestion. Thank you.

simonguinot reviewed (commented):
Thanks @soburi for taking care of this cleaning. See my comments below.

soburi committed:
boards: seagate: legend: Cleanup `led-strip` alias

The led-strip alias can be determined in the base dts file.
Remove from overlay files and aggregate to the base dts file.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

erwango reviewed (commented):
Approved on my side, now requires assignees approval.

simonguinot commented on samples/drivers/led_strip/src/main.c:
```diff
@@ -31,7 +39,7 @@ static const struct led_rgb colors[] = {
+static struct led_rgb pixels[STRIP_NUM_PIXELS];
 
 static const struct device *const strip = DEVICE_DT_GET(STRIP_NODE);
 
```
While you are at it, please remove the new line below (at line 63), and eventually add one the `led_strip_update_rgb()` call.

And eventually shortened the `color` computation to a single line.:

`color = (color + 1) % ARRAY_SIZE(colors);`

soburi replied:
Improved loop handling to make it easier to understand.

simonguinot replied:
It is much better, thanks !

simonguinot commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -8,36 +8,47 @@
 	select PICOSDK_USE_PIO
 	help
-	  Use the PIO feature available on RaspberryPi Pico devices.
+	  Enable driver for WS2812 (and compatibles) LED strip controller using
```
Please use the same wording in all the messages: "LED strip controller" or "LED strip", 

Sorry for the nitpicking.

soburi replied:
I revised the message a little.

simonguinot reviewed (commented):
Some nitpicks :)

simonguinot commented:
> Approved on my side, now requires assignees approval.

Hi @erwango. It looks like you didn't actually approve the PR.

simonguinot reviewed (approved):
Thanks @soburi for this patch series cleaning the led_strip subsystem !

soburi commented:
@thedjnK 
Could you take a look, please?

soburi commented:
@thedjnK 
ping

kartben commented:
@soburi thanks for the samples cleanup! Much appreciated.
Could you please make sure there is a proper redirect in `doc/_scripts/redirects.py` so that the links "The latest development version of this page may be more current than this released 3.6.0 version." in e.g. https://docs.zephyrproject.org/3.6.0/samples/drivers/led_apa102/README.html work propertly? 

Thanks!

kartben commented:
@soburi I think this change should be described in the migration guide. Just had an application from a few months ago that failed to compile as the conf overlay was defining `CONFIG_WS2812_STRIP=y`
